packages/aws-rfdk/lib/lambdas/nodejs/pad-efs-storage/handlers.ts

```
@@ -0,0 +1,192 @@
+ * @param numFilesToAdd How many numbered files to add.
+ * @param filesize Size, in MB, of the files to create.
+ * @param mountPoint Directory in which to create the directory.
+ */
+export async function growFilesystem(numFilesToAdd: number, filesize: number, mountPoint: string): Promise<void> {
+  // We find the highest numbered file created thus far, and start adding numbered files after it.
+  var filename: string = await determineNextSequentialFilename(mountPoint);
+  for (var i=0; i<numFilesToAdd; i++) {
+    const outfilename = join(mountPoint, filename);
+    await writePaddingFile(outfilename, filesize);
+    filename = nextSequentialFile(filename);
+  }
+}
+
+/**
+ * Delete a given number of numbered files from the given filesystem.
+ * Note: exported so that we can test it.
+ * @param numFilesToRemove How many files to remove from the directory.
+ * @param mountPoint Directory from which to delete files.
+ */
+export async function shrinkFilesystem(numFilesToRemove: number, mountPoint: string): Promise<void> {
+  // Find all of the numbered "files" in the directory, and then delete the highest numbered ones until
+  // we've deleted as many as we need to.
+  const numberedFiles = await listNumberedFiles(mountPoint);
+  var numFilesDeleted = 0;
+  var index = numberedFiles.length - 1;
+  while (numFilesDeleted < numFilesToRemove && index >= 0) {
+    const filename = join(mountPoint, numberedFiles[index]);
+    try {
+      const stat = await fsp.stat(filename);
+      if (stat.isFile()) {
+        console.log(`Deleting: ${filename}`);
+        try {
+          await fsp.unlink(filename);
+          numFilesDeleted += 1;
+        } catch (err) {
+          console.error(`Unable to delete: ${filename} -- Error: ${err.message}`);
+        }
+      }
+    } catch (err) {
+      console.warn(`Warning: Unable to stat file '${filename}'`);
+    }
+    index -= 1;
+  }
+  console.log(`Deleted ${numFilesDeleted} numbered files`);
+}
```

It looks like errors are never surfaced to the caller here. I understand we want to be robust here and attempt to skip a file if we encounter errors and try deleting other files.

Can we change the code to throw an error at the end if we were unable to delete the desired number of files by the end of the loop? That way the state machine would actually show an error for this state and transition to the fail state.

---

Resolved